WordPress/src/main/java/org/wordpress/android/ui/posts/PostPreviewActivity.java

```diff
@@ -116,17 +158,142 @@ public boolean onOptionsItemSelected(final MenuItem item) {
+
+        if (mPost == null
+                || mIsUpdatingPost
+                || PostUploadService.isPostUploading(mPost.getLocalTablePostId())
+                || (!mPost.isLocalChange() && !mPost.isLocalDraft())) {
+            messageView.setVisibility(View.GONE);
+            return;
+        }
+
+        TextView messageText = (TextView) messageView.findViewById(R.id.message_text);
+        if (mPost.isLocalChange()) {
+            messageText.setText(R.string.local_changes_explainer);
+        } else if (mPost.isLocalDraft()) {
+            messageText.setText(R.string.local_draft_explainer);
+        }
+
+        // publish applies to both local draft and local changes
+        View btnPublish = messageView.findViewById(R.id.btn_publish);
+        btnPublish.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                AniUtils.animateBottomBar(messageView, false);
+                publishPost();
+            }
+        });
+
+        // revert applies to only local changes
+        View btnRevert = messageView.findViewById(R.id.btn_revert);
+        btnRevert.setVisibility(mPost.isLocalDraft() ? View.GONE : View.VISIBLE);
+        if (!mPost.isLocalDraft()) {
+            btnRevert.setOnClickListener(new View.OnClickListener() {
+                @Override
+                public void onClick(View v) {
+                    AniUtils.animateBottomBar(messageView, false);
+                    revertPost();
+                }
+            });
+        }
+
+        // first set message bar to invisible so it takes up space, then animate it in
+        // after a brief delay to give time for preview to render first
+        messageView.setVisibility(View.INVISIBLE);
+        new Handler().postDelayed(new Runnable() {
+            @Override
+            public void run() {
+                if (!isFinishing() && messageView.getVisibility() != View.VISIBLE) {
+                    AniUtils.animateBottomBar(messageView, true);
+                }
+            }
+        }, 1000);
+    }
+
+    /*
+     * reverts local changes for this post, replacing it with the latest version from the server
+     */
+    private void revertPost() {
+        if (isFinishing() || !NetworkUtils.checkConnection(this)) {
+            return;
+        }
+
+        if (mIsUpdatingPost) {
+            AppLog.d(AppLog.T.POSTS, "post preview > already updating post");
+        } else {
+            new UpdatePostTask().executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR);
+        }
+    }
+
+    private void publishPost() {
+        if (!isFinishing() && NetworkUtils.checkConnection(this)) {
+            PostUploadService.addPostToUpload(mPost);
+            startService(new Intent(this, PostUploadService.class));
+        }
+    }
+
+    @SuppressWarnings("unused")
+    public void onEventMainThread(PostEvents.PostUploadStarted event) {
```

Comment:
@daniloercoli Are you referring to the two `onEventMainThread` methods? If so, those are necessary in order to update the UI when the post upload has started and ended. `UpdatePostTask` is entirely separate - it simply gets the latest version of the post.